File: framework/backend/kotlin/packages/k1s0-consensus/src/main/kotlin/dev/k1s0/consensus/SagaResult.kt
package dev.k1s0.consensus

import java.time.Instant

/**
 * Status of a saga execution.
 */
public enum class SagaStatus {
    /** The saga is currently executing steps. */
    RUNNING,

    /** All steps completed successfully. */
    COMPLETED,

    /** A step failed and compensation is in progress. */
    COMPENSATING,

    /** Compensation completed after a failure. */
    COMPENSATED,

    /** The saga failed and could not be compensated. */
    FAILED,

    /** The saga has been moved to the dead letter queue. */
    DEAD_LETTER,
}

/**
 * Persistent representation of a saga instance.
 *
 * @property sagaId Unique identifier for this saga execution.
 * @property sagaName The name of the saga definition.
 * @property status Current execution status.
 * @property currentStep The index of the current or failed step.
 * @property context Serialized context data.
 * @property error Error message if the saga failed.
 * @property retryCount Number of retries attempted.
 * @property createdAt When the saga was created.
 * @property updatedAt When the saga was last updated.
 */
public data class SagaInstance(
    val sagaId: String,
    val sagaName: String,
    val status: SagaStatus,
    val currentStep: Int,
    val context: String,
    val error: String? = null,
    val retryCount: Int = 0,
    val createdAt: Instant,
    val updatedAt: Instant,
)

/**
 * Result of a saga execution.
 *
 * @property sagaId Unique identifier for this saga execution.
 * @property status Final status of the saga.
 * @property context The final context after execution.
 * @property error Error message if the saga failed.
 * @property completedSteps List of step names that completed successfully.
 * @property compensatedSteps List of step names that were compensated.
 */
public data class SagaResult(
    val sagaId: String,
    val status: SagaStatus,
    val context: Map<String, Any>,
    val error: String? = null,
    val completedSteps: List<String> = emptyList(),
    val compensatedSteps: List<String> = emptyList(),
)

File: framework/backend/csharp/src/K1s0.Consensus/SagaResult.cs
using System.Text.Json;

namespace K1s0.Consensus;

/// <summary>
/// Status of a saga instance.
/// </summary>
public enum SagaStatus
{
    /// <summary>
    /// Saga is currently executing steps.
    /// </summary>
    Running,

    /// <summary>
    /// All steps completed successfully.
    /// </summary>
    Completed,

    /// <summary>
    /// A step failed and compensation was executed.
    /// </summary>
    Compensated,

    /// <summary>
    /// A step failed and compensation also failed. Requires manual intervention.
    /// </summary>
    Failed,

    /// <summary>
    /// Saga is paused, awaiting resume.
    /// </summary>
    Suspended
}

/// <summary>
/// The result of a saga execution.
/// </summary>
/// <typeparam name="TContext">The saga context type.</typeparam>
public sealed class SagaResult<TContext> where TContext : class
{
    /// <summary>
    /// The final status of the saga.
    /// </summary>
    public SagaStatus Status { get; init; }

    /// <summary>
    /// The final context state.
    /// </summary>
    public TContext? Context { get; init; }

    /// <summary>
    /// The saga instance identifier.
    /// </summary>
    public string SagaId { get; init; } = string.Empty;

    /// <summary>
    /// Errors encountered during execution or compensation, keyed by step name.
    /// </summary>
    public IReadOnlyDictionary<string, string> Errors { get; init; } = new Dictionary<string, string>();

    /// <summary>
    /// The index of the last successfully completed step (0-based), or -1 if none completed.
    /// </summary>
    public int LastCompletedStep { get; init; } = -1;
}

/// <summary>
/// A persisted saga instance record.
/// </summary>
public sealed class SagaInstance
{
    /// <summary>
    /// Unique identifier for this saga instance.
    /// </summary>
    public string SagaId { get; init; } = string.Empty;

    /// <summary>
    /// The saga definition name.
    /// </summary>
    public string SagaName { get; init; } = string.Empty;

    /// <summary>
    /// Current status.
    /// </summary>
    public SagaStatus Status { get; set; }

    /// <summary>
    /// The serialized context JSON.
    /// </summary>
    public string ContextJson { get; set; } = "{}";

    /// <summary>
    /// The index of the current step being executed.
    /// </summary>
    public int CurrentStep { get; set; }

    /// <summary>
    /// When this instance was created.
    /// </summary>
    public DateTimeOffset CreatedAt { get; init; } = DateTimeOffset.UtcNow;

    /// <summary>
    /// When this instance was last updated.
    /// </summary>
    public DateTimeOffset UpdatedAt { get; set; } = DateTimeOffset.UtcNow;

    /// <summary>
    /// Serialized error information as JSON.
    /// </summary>
    public string ErrorsJson { get; set; } = "{}";

    /// <summary>
    /// Deserializes the context from JSON.
    /// </summary>
    /// <typeparam name="TContext">The context type.</typeparam>
    /// <returns>The deserialized context.</returns>
    public TContext? DeserializeContext<TContext>() where TContext : class =>
        JsonSerializer.Deserialize<TContext>(ContextJson);
}
